src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,207 @@
+*/
+public class PostThumbnailsHandler extends AbstractFesterHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PostThumbnailsHandler.class, Constants.MESSAGES);
+
+    private static final String MANIFESTS = "manifests";
+
+    private static final String ATTACHMENT = "attachment; filename=\"{}\"";
+
+    private static final String BR_TAG = "<br>";
+
+    private static final String ID_TAG = "@id";
+
+    private final String myExceptionPage;
+
+    private final String myFesterizeVersion;
+
+    /**
+     * FUA = Festerize User Agent
+     */
+    private final Pattern myFUAPattern;
+
+    /**
+     * @param aVertx
+     * @param aConfig
+    */
+    public PostThumbnailsHandler(final Vertx aVertx, final JsonObject aConfig) throws IOException {
+        super(aVertx, aConfig);
+
+        final byte[] bytes = IOUtils.readBytes(getClass().getResourceAsStream("/webroot/error.html"));
+
+        myExceptionPage = new String(bytes, StandardCharsets.UTF_8);
+        myFesterizeVersion = aConfig.getString(Config.FESTERIZE_VERSION);
+        myFUAPattern = Pattern.compile("Festerize/(?<version>\\d+\\.\\d+\\.\\d+)");
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see io.vertx.core.Handler#handle(java.lang.Object)
+     */
+    @Override
+    public void handle(final RoutingContext aContext) {
+        final HttpServerRequest request = aContext.request();
+        final HttpServerResponse response = aContext.response();
+        final Set<FileUpload> csvUploads = aContext.fileUploads();
+        final String festerizeUserAgent = StringUtils.trimTo(request.getHeader("User-Agent"), Constants.EMPTY);
+        final Matcher festerizeUserAgentMatcher = myFUAPattern.matcher(festerizeUserAgent);
+        final String errorMessage;
+
+        // An uploaded CSV is required
+        if (csvUploads.isEmpty()) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_037);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else if (festerizeUserAgentMatcher.matches() &&
+                   !festerizeUserAgentMatcher.group("version").equals(myFesterizeVersion)) {
+            errorMessage = LOGGER.getMessage(MessageCodes.MFS_147, myFesterizeVersion);
+            returnError(response, HTTP.BAD_REQUEST, errorMessage);
+        } else {
+            final FileUpload csvFile = csvUploads.iterator().next();
+            final String filePath = csvFile.uploadedFileName();
+            final String fileName = csvFile.fileName();
+
+            try {
+                final List<String[]> originalLines =
+                    new CSVReader(Files.newBufferedReader(Paths.get(filePath))).readAll();
+                final List<String[]> linesWithThumbs = ThumbnailUtils.addThumbnailColumn(originalLines);
+                final int manifestIndex = Arrays.asList(linesWithThumbs.get(0)).indexOf(CSV.MANIFEST_URL);
+                final String collectionURL = linesWithThumbs.get(1)[manifestIndex];
```

Comment:
The first line of a CSV is not necessarily a collection row (and I don't believe we are concerned with thumbnails of Collections, anyway) -- you can use `CsvParser.getObjectType` to determine the object type represented by the row

Reply:
Yes, good point. Another way to approach the problem would be to get the works from the CSV and query them for their canvases, skipping the collection altogether.

Reply:
agreed, best to just skip collection rows